src/hb-ot-shape-math.cc

```diff
@@ -0,0 +1,717 @@
+          overlapDelta /= 2;
+        }
+
+        if (overlapDelta > 0) {
+          deltaSum += overlapDelta;
+          extraSize -= 2 * overlapDelta;
+        }
+      }
+
+      // Update the position of the glyphs.
+      if (glyphAssembly.is_horizontal()) {
+        buffer->pos[glyphIndexTowardsStart].x_offset += deltaSum;
+        buffer->pos[glyphIndexTowardsEnd].x_offset -= deltaSum;
+      } else {
+        buffer->pos[glyphIndexTowardsStart].y_offset += deltaSum;
+        buffer->pos[glyphIndexTowardsEnd].y_offset -= deltaSum;
+      }
+
+      glyphIndexTowardsStart--;
+      glyphIndexTowardsEnd++;
+    }
+  }
+
+  // Adjust the advance for the glyph assembly.
+  hb_position_t max_orthogonal_advance =
+    get_glyph_assembly_max_orthogonal_advance (font, glyphAssembly);
+  if (glyphAssembly.is_horizontal()) {
+    buffer->pos[glyphCount - 1].x_advance =
+      buffer->pos[glyphCount - 1].x_offset -
+      buffer->pos[0].x_offset +
+      glyphAssembly.full_advance(buffer->info[glyphCount - 1].cluster);
+    buffer->pos[glyphCount - 1].y_advance = -max_orthogonal_advance;
+  } else {
+    buffer->pos[glyphCount - 1].x_advance = max_orthogonal_advance;
+    buffer->pos[glyphCount - 1].y_advance =
+      -(buffer->pos[glyphCount - 1].y_offset -
+        buffer->pos[0].y_offset +
+        glyphAssembly.full_advance(buffer->info[glyphCount - 1].cluster));
+  }
+
+  // We now move all parts into the same cluster and shift the assembly to
+  // ensure that the top left glyph is located at the origin.
+  hb_position_t delta = glyphAssembly.is_horizontal() ?
+    -buffer->pos[0].x_offset : -buffer->pos[glyphCount - 1].y_offset;
+  for (unsigned int j = 0; j < glyphCount; j++) {
+    buffer->info[j].cluster = initialCluster;
+    if (glyphAssembly.is_horizontal())
+      buffer->pos[j].x_offset += delta;
+    else
+      buffer->pos[j].y_offset += delta;
+  }
+
+  return true;
+}
+
+/**
+ * hb_ot_shape_math_stretchy:
+ * @font: an #hb_font_t a math font
+ * @buffer: an #hb_buffer_t to stretch, containing only a single element
+ * @horizontal: boolean indicating the stretch direction
+ * @target_size: minimal size to which the glyph should be stretched
+ *
+ * Try and stretch the single element of @buffer to a specified target size.
+ * If the content type of the buffer is HB_BUFFER_CONTENT_TYPE_UNICODE then that
+ * element is a Unicode character. The stretching is performed using only
+ * predefined Unicode glyph assembly such as
+ *
+ *                              |LEFT PARENTHESIS UPPER HOOK (U+239B)
+ * LEFT PARENTHESIS (U+0028) => |LEFT PARENTHESIS EXTENSION  (U+239C)
+ *                              |LEFT PARENTHESIS LOWER HOOK (U+239D)
+ *
+ * If the content type of the buffer is HB_BUFFER_CONTENT_TYPE_GLYPHS then that
+ * element is a glyph index and we try and use the OpenType MATH table to
+ * find a larger glyph from a MathGlyphVariantRecord subtable or a glyph
+ * assembly described in a GlyphAssembly subtable.
```

Comment:
I’m thinking this wouldn’t be much useful, since I expect the clients to check of the fonts has a math table before using any of these new API calls, and the chance that a font has a MATH table without such basic stretchy glyphs is unlikely and if it does it is a font bug after all. On the other hand clients might intentionally call this method without checking for MATH table to get the fallback, and HarfBuzz does fallback text shaping as well so this is not something new. But why stop here and don’t provide fallback math constants as well?